File: Testing/Source/DataStructureAndEncodingDefinition/Cxx/TestReaderUpToTag.cxx
/*=========================================================================

  Program: GDCM (Grassroots DICOM). A DICOM library
  Module:  $URL$

  Copyright (c) 2006-2010 Mathieu Malaterre
  All rights reserved.
  See Copyright.txt or http://gdcm.sourceforge.net/Copyright.html for details.

     This software is distributed WITHOUT ANY WARRANTY; without even
     the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
     PURPOSE.  See the above copyright notice for more information.

=========================================================================*/
#include "gdcmReader.h"
#include "gdcmFileMetaInformation.h"
#include "gdcmFile.h"
#include "gdcmTesting.h"
#include "gdcmMediaStorage.h"


int TestReadUpToTag(const char* filename, bool verbose = false)
{
  if( verbose )
  std::cout << "TestRead: " << filename << std::endl;

  std::ifstream is( filename );
  gdcm::Reader reader;
  reader.SetStream( is );
  // Let's read up to Pixel Data el...
  gdcm::Tag pixeldata (0x7fe0,0x0010);
  std::set<gdcm::Tag> skiptags;
  // ... but do not read it (to skip mem allocation)
  skiptags.insert( pixeldata );
  if ( !reader.ReadUpToTag( pixeldata, skiptags) )
    {
    std::cerr << "TestReadError: Failed to read: " << filename << std::endl;
    return 1;
    }
  std::streamoff outStreamOffset = is.tellg();

  const gdcm::FileMetaInformation &h = reader.GetFile().GetHeader();

  const gdcm::DataSet &ds = reader.GetFile().GetDataSet();

  if(verbose)
    std::cout << "{ \"" << filename << "\"," << outStreamOffset << " }," << std::endl;
  std::streamoff refoffset = gdcm::Testing::GetStreamOffsetFromFile(filename);
  if( refoffset != outStreamOffset )
    {
    std::cerr << filename << ": " << outStreamOffset << " should be " << refoffset << std::endl;
    return 1;
    }
  is.close();

  return 0;
}

int TestReaderUpToTag(int argc, char *argv[])
{
  if( argc == 2 )
    {
    const char *filename = argv[1];
    return TestReadUpToTag(filename, true);
    }

  // else
  gdcm::Trace::DebugOff();
  gdcm::Trace::WarningOff();
  int r = 0, i = 0;
  const char *filename;
  const char * const *filenames = gdcm::Testing::GetFileNames();
  while( (filename = filenames[i]) )
    {
    r += TestReadUpToTag( filename );
    ++i;
    }

  return r;
}

File: Testing/Source/DataStructureAndEncodingDefinition/Java/TestReaderUpToTag.java
/*=========================================================================

  Program: GDCM (Grassroots DICOM). A DICOM library

  Copyright (c) 2006-2011 Mathieu Malaterre
  All rights reserved.
  See Copyright.txt or http://gdcm.sourceforge.net/Copyright.html for details.

     This software is distributed WITHOUT ANY WARRANTY; without even
     the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
     PURPOSE.  See the above copyright notice for more information.

=========================================================================*/
/*
 */
import gdcm.*;
import java.util.SortedSet;
import java.util.TreeSet;

public class TestReaderUpToTag
{
  public static void main(String[] args) throws Exception {

    long nfiles = Testing.GetNumberOfFileNames();

    Tag tags[] = {
      new Tag(0x8,0x8),
      new Tag(0x10,0x10) };
    SortedSet<Tag> tagSet = new TreeSet<Tag>();
    for( Tag tag : tags ) {
      tagSet.add( tag );
    }
    Tag last = tagSet.last();
    System.out.println("last: " + last);
    TagSetType skip = new TagSetType();
    Tag pixelData = new Tag(0x7fe0,0x0010);
    skip.insert( pixelData );
    Trace.DebugOff();
    Trace.WarningOff();
    StringFilter sf = new StringFilter();
    for( long i = 0; i < nfiles; ++i ) {
      String filename = Testing.GetFileName( i );
      if( filename.contains( "ExplicitVRforPublicElementsImplicitVRforShadowElements" )
          || filename.contains( "SIEMENS_SOMATOM-12-ACR_NEMA-ZeroLengthUs" )
        ) {
        Reader reader = new Reader();
        reader.SetFileName( filename );
        if ( !reader.ReadUpToTag(pixelData, skip) )
        {
          throw new Exception("Could not read: " + filename );
        }
        System.out.println("reading: " + filename);
        File file = reader.GetFile();
        sf.SetFile( file );
        DataSet ds = file.GetDataSet();
        for( Tag tag : tags ) {
          if( ds.FindDataElement( tag ) ) {
            DataElement de = ds.GetDataElement( tag );
            System.out.println("de: " + tag + " -> " + sf.ToString( de ) );
          }
        }
      }
    }
  }
}
